Allow lists and dicts as msg/var parameters to debug module

gnosek opened the pull request:
##### SUMMARY
This pull request allows lists and dicts to be passed as values of the `debug` module. Strictly speaking, a similar effect may be achieved using `with_items`, but the output is clearer when implemented natively. Please see additional information for before/after comparison:

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
`plugins/action/debug.py`

##### ANSIBLE VERSION
```
ansible 2.5.0 (debug-list 4b50d22be3) last updated 2017/10/14 12:09:47 (GMT +200)
  config file = None
  configured module search path = [u'/Users/blackfire/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/blackfire/src/ansible/lib/ansible
  executable location = /Users/blackfire/src/ansible/bin/ansible
  python version = 2.7.10 (default, Jul 14 2015, 19:46:27) [GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)]
```


##### ADDITIONAL INFORMATION

Before:
```
  - debug:
      var: "{{ item }}"
    with_items:
    - str1_var
    - str2_var
```
```
ok: [localhost] => (item=str1_var) => {
    "item": "str1_var",
    "str1_var": "string variable 1"
}
ok: [localhost] => (item=str2_var) => {
    "item": "str2_var",
    "str2_var": "string variable 2"
}
```

After:
```
  - debug:
      var:
      - str1_var
      - str2_var
```
```
ok: [localhost] => {
    "var": {
        "str1_var": "string variable 1",
        "str2_var": "string variable 2"
    }
}
```

gnosek committed:
[debug] Extract `var` parameter to a separate variable

Just a simple cleanup

gnosek committed:
[debug] Use type.__name__ for more readable output

Before:
    "<type 'list'>": "VARIABLE IS NOT DEFINED!"

After:
    "list": "VARIABLE IS NOT DEFINED!"

gnosek committed:
[debug] Support lists/dicts in debug var parameter

Now the following construct is legal:

    - debug:
        var:
        - str1_var
        - str2_var

It templates all list items separately and returns them as a dict:

    ok: [localhost] => {
        "list": {
            "str1_var": "string variable 1",
            "str2_var": "string variable 2"
        }
    }

It works the same way for dicts:

    - debug:
        var:
          str1_var_label: str1_var
          str2_var_label: str2_var

Which expands into:

    ok: [localhost] => {
        "dict": {
            "str1_var_label": "string variable 1",
            "str2_var_label": "string variable 2"
        }
    }

gnosek committed:
[debug] Document lists/dicts as debug module parameters

gnosek committed:
[debug] DRY: Extract usage of the templar into a lambda

gnosek committed:
Fix ansible-test sanity checks

failure: The test `ansible-test sanity --test pep8` failed with the following error:

lib/ansible/plugins/action/debug.py:58:21: E731 do not assign a lambda expression, use a def

gnosek committed:
[debug] Remove one nesting level in return value

Instead of returning a dict called 'list' or 'dict',
return the debug values at the top level of the result

gnosek committed:
[debug] Use consistent logic for scalars and list/dict items

Now referring to a nonexistent variable in a list/dict yields:

    ok: [localhost] => {
        "str1_var": "string variable 1",
        "str2_var": "string variable 2",
        "str3_var": "VARIABLE IS NOT DEFINED!"
    }

dagwieers reviewed (approved):
I like this functionality, but leave it up to core developers to merge this.

gnosek commented:
Hmm,

The test failures at https://app.shippable.com/github/ansible/ansible/runs/40890/33/tests show usage like `debug var={{ external_ip }}` -- this doesn't seem legal (it would evaluate to e.g. `debug var=10.0.0.1`) but if it worked before, it's probably widespread enough to cause compatibility problems. I'll see what I can do about it

gnosek committed:
Revert "[debug] Remove one nesting level in return value"

This reverts commit c2ac76457063ac16eb5a76a3442ad4b61af52b55.

With `debug: var={{ a_dict_variable }}` we would have overwritten
all variables matching keys of the variable. While this isn't really
proper syntax, it's popular enough to end up in the Ansible test
suite.

gnosek committed:
Don't use debug: var={{ ... }} in tests

This has probably never worked (tried to access a weird named variable).

bcoca commented:
@gnosek that test was incorrect, but the expression is legal, you should not fail if var points to a template.

gnosek commented:
@bcoca, the tests didn't fail. I had found the error you mention earlier and fixed it by reverting one commit. AFAIK the failures come from an OpenSUSE repo having problems. I fixed the tests as a cleanup but if you prefer to keep them as is, I can drop that commit.

mkrizek commented:
@gnosek are you still interested in this PR? If so, would you mind rebasing and resolving conflicts? Thanks!

Akasurde commented:
@gnosek Are you still working on this? Thanks.

needs_info
